core/trie/bitarray.go

```diff
@@ -0,0 +1,458 @@
+		// e.g., if len=200, keeps lowest 8 bits (200-192)
+		mask := maxUint64 >> (256 - uint16(b.len))
+		binary.BigEndian.PutUint64(res[0:8], b.words[3]&mask)
+		binary.BigEndian.PutUint64(res[8:16], b.words[2])
+		binary.BigEndian.PutUint64(res[16:24], b.words[1])
+		binary.BigEndian.PutUint64(res[24:32], b.words[0])
+	case b.len >= 128:
+		// Mask for bits 128-191: keeps only valid bits above 128
+		// e.g., if len=150, keeps lowest 22 bits (150-128)
+		mask := maxUint64 >> (192 - b.len)
+		binary.BigEndian.PutUint64(res[8:16], b.words[2]&mask)
+		binary.BigEndian.PutUint64(res[16:24], b.words[1])
+		binary.BigEndian.PutUint64(res[24:32], b.words[0])
+	case b.len >= 64:
+		// You get the idea
+		mask := maxUint64 >> (128 - b.len)
+		binary.BigEndian.PutUint64(res[16:24], b.words[1]&mask)
+		binary.BigEndian.PutUint64(res[24:32], b.words[0])
+	default:
+		mask := maxUint64 >> (64 - b.len)
+		binary.BigEndian.PutUint64(res[24:32], b.words[0]&mask)
+	}
+
+	return res[:]
+}
+
+// Sets b to the least significant 'n' bits of x.
+// If n >= x.len, b is an exact copy of x.
+// For example:
+//
+//	x = 11001011 (len=8)
+//	LSBs(x, 4) = 1011 (len=4)
+//	LSBs(x, 10) = 11001011 (len=8, original x)
+//	LSBs(x, 0) = 0 (len=0)
+//
+//nolint:mnd
+func (b *BitArray) LSBs(x *BitArray, length uint8) *BitArray {
+	if length >= x.len {
+		return b.Set(x)
+	}
+
+	b.Set(x)
+	b.len = length
+
+	// Clear all words beyond what's needed
+	switch {
+	case length == 0:
+		b.words = [4]uint64{0, 0, 0, 0}
+	case length <= 64:
+		mask := maxUint64 >> (64 - length)
+		b.words[0] &= mask
+		b.words[1] = 0
+		b.words[2] = 0
+		b.words[3] = 0
+	case length <= 128:
+		mask := maxUint64 >> (128 - length)
+		b.words[1] &= mask
+		b.words[2] = 0
+		b.words[3] = 0
+	case length <= 192:
+		mask := maxUint64 >> (192 - length)
+		b.words[2] &= mask
+		b.words[3] = 0
+	default:
+		mask := maxUint64 >> (256 - uint16(length))
+		b.words[3] &= mask
+	}
+
+	return b
+}
+
+// Checks if the current bit array share the same most significant bits with another, where the length of
+// the check is determined by the shorter array. Returns true if either array has
+// length 0, or if the first min(b.len, x.len) MSBs are identical.
+//
+// For example:
+//
+//	a = 1101 (len=4)
+//	b = 11010111 (len=8)
+//	a.EqualMSBs(b) = true  // First 4 MSBs match
+//
+//	a = 1100 (len=4)
+//	b = 1101 (len=4)
+//	a.EqualMSBs(b) = false // All bits compared, not equal
+//
+//	a = 1100 (len=4)
+//	b = [] (len=0)
+//	a.EqualMSBs(b) = true  // Zero length is always a prefix match
+func (b *BitArray) EqualMSBs(x *BitArray) bool {
+	if b.len == x.len {
+		return b.Equal(x)
+	}
+
+	if b.len == 0 || x.len == 0 {
+		return true
+	}
+
+	// Compare only the first min(b.len, x.len) bits
+	minLen := b.len
+	if x.len < minLen {
+		minLen = x.len
+	}
+
+	return new(BitArray).MSBs(b, minLen).Equal(new(BitArray).MSBs(x, minLen))
+}
+
+// Sets b to the most significant 'n' bits of x.
```

Comment:
```suggestion
// Sets the bit array to the most significant 'n' bits of x.
```
Suggesting this approach because someone reading the description might be unaware that `b` refers to the bit array itself